### PR TITLE
chore(cli): use option source name for deprecation warnings

### LINKED ISSUE #15568: Unhelpful warning is displayed when using deprecated configs
When a deprecated setting is used, a warning is posted but it doesn't indicate the flag or the env to be used, but rather its name.

`PrintDeprecatedOptions()` is where this is defined. We should probably print the CLI flag, env var name, or a link to the documentation.

```
WARN: Notifications: Email Auth: Password is deprecated, please use Email Auth: Password instead.
WARN: Telemetry (backwards compatibility) is deprecated, please use Telemetry Enable instead.
```

### PR DESCRIPTION
**Pull request by ethanndickson**
Closes #15568.

This improves the deprecation warnings, so that they use the config name for however they were supplied for both the deprecated option, and the recommended options.

(I'm assuming it's relatively obvious for users this is an env variable)
```
$ set -x CODER_ADDRESS 127.0.0.1:3003
$ coder server
WARN: `CODER_ADDRESS` is deprecated, please use `CODER_HTTP_ADDRESS` and `CODER_TLS_ADDRESS` instead.
```

```
$ coder server --address 127.0.0.1:3003
WARN: `--address` is deprecated, please use `--http-address` and `--tls-address` instead.
```
YAML names are fully qualified, to make it clear they're set via YAML
```
$ coder server --config config.yaml
WARN: `introspection.logging.verbose` is deprecated, please use `introspection.logging.filter` instead.
```

This PR also applies the middleware to *all* commands.

### CODE REVIEW

**Comment by ethanndickson** [comment]
* **#15581** <a href="https://app.graphite.dev/github/pr/coder/coder/15581?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/coder/coder/15581?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: cli/root.go [hunk 1]
```diff
@@ -1307,3 +1307,53 @@ func headerTransport(ctx context.Context, serverURL *url.URL, header []string, h
...
+			for _, opt := range opts {
+				if opt.UseInstead == nil {
+					continue
+				}
+
+				if opt.ValueSource == serpent.ValueSourceNone || opt.ValueSource == serpent.ValueSourceDefault {
+					continue
+				}
+
+				warnStr := translateSource(opt.ValueSource, opt) + " is deprecated, please use "
```
Comment (dannykopping) [inline cli/root.go:1327]: Nit: this could be a good place to use a `strings.Builder`.

File: cli/root.go [hunk 2]
```diff
@@ -1307,3 +1307,53 @@ func headerTransport(ctx context.Context, serverURL *url.URL, header []string, h
...
+// translateSource provides the name of the source of the option, depending on the
+// supplied target ValueSource.
+func translateSource(target serpent.ValueSource, opt serpent.Option) string {
+	switch target {
+	case serpent.ValueSourceFlag:
+		return fmt.Sprintf("`--%s`", opt.Flag)
+	case serpent.ValueSourceEnv:
+		return fmt.Sprintf("`%s`", opt.Env)
+	case serpent.ValueSourceYAML:
+		return fmt.Sprintf("`%s`", opt.YAML)
```
Comment (dannykopping) [inline cli/root.go:1355]: I worry that YAML options won't read super well; we may need to show the full "address" of this option.

File: cli/restart.go [hunk 3]
```diff
@@ -26,6 +26,7 @@ func (r *RootCmd) restart() *serpent.Command {
 		Short:       "Restart a workspace",
 		Middleware: serpent.Chain(
 			serpent.RequireNArgs(1),
+			PrintDeprecatedOptions(),
```
Comment (dannykopping) [inline cli/restart.go:29]: Good idea to use middleware 👍 
I'm curious why this isn't applied to all CLI actions, though?

Serpent may not have middleware support but we could centralize the definition of the common middlewares; i.e. (almost?) every CLI action needs to call `r.InitClient(client)` and will need this middleware, so let's define them centrally.

**Review by dannykopping — COMMENTED** [review]
Thanks for picking this up @ethanndickson!

File: cli/root.go (hunk 2 above)
Comment (ethanndickson) [reply cli/root.go:1355]: Good idea!
```
$ coder server --config config.yaml
WARN: `introspection.logging.verbose` is deprecated, please use `introspection.logging.filter` instead.
```

File: cli/root.go [hunk 4]
```diff
@@ -1324,17 +1333,20 @@ func PrintDeprecatedOptions() serpent.MiddlewareFunc {
 					continue
 				}
 
-				warnStr := translateSource(opt.ValueSource, opt) + " is deprecated, please use "
+				var warnStr strings.Builder
+				warnStr.Cap()
```
Comment (dannykopping) [inline cli/root.go:1337]: Was this left in by accident, perhaps?

```suggestion
```

**Review by dannykopping — APPROVED** [review]
Great work!

File: cli/root.go (hunk 4 above)
Comment (ethanndickson) [reply cli/root.go:1337]: oops, thanks!

**Pull request merged**